rclcpp/src/rclcpp/executors/detail/timer_manager.hpp

```diff
@@ -0,0 +1,356 @@
+    }
+
+    if (clock_type_of_timer->type != timer_type) {
+      // timer is handled by another queue
+      return;
+    }
+
+    std::unique_ptr<TimerData> data = std::make_unique<TimerData>();
+    data->timer_ready_callback = std::move(timer_ready_callback);
+    data->rcl_ref = std::move(handle);
+
+    timer->set_on_reset_callback(
+      [data_ptr = data.get(), this](size_t) {
+        std::scoped_lock l(mutex);
+        if (!remove_if_dropped(data_ptr))
+        {
+          add_timer_to_running_map(data_ptr);
+        }
+      });
+
+    {
+      std::scoped_lock l(mutex);
+      add_timer_to_running_map(data.get());
+
+      all_timers.emplace_back(std::move(data) );
+    }
+
+    //wake up thread as new timer was added
+    used_clock_for_timers.cancel_sleep_or_wait();
+  }
+
+private:
+  /**
+   * Checks if the timer is still referenced if not deletes it from the queue
+   *
+   * @param timer_data The timer to check
+   * @return true if removed / invalid
+   */
+  bool remove_if_dropped(const TimerData * timer_data)
+  {
+    if (timer_data->rcl_ref.unique()) {
+      // timer was deleted
+      auto it = std::find_if(
+        all_timers.begin(), all_timers.end(), [timer_data](const std::unique_ptr<TimerData> & e) {
+          return timer_data == e.get();
+        }
+      );
+
+      if (it != all_timers.end()) {
+        all_timers.erase(it);
+      }
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @brief adds the given timer_data to the map of running timers, if valid.
+   *
+   * Advances the rcl timer.
+   * Computes the next call time of the timer.
+   * readds the timer to the map of running timers
+   */
+  void add_timer_to_running_map(const TimerData * timer_data)
+  {
+    rcl_ret_t ret = rcl_timer_call(const_cast<rcl_timer_t *>(timer_data->rcl_ref.get()));
+    if (ret == RCL_RET_TIMER_CANCELED) {
+      return;
+    }
+
+    int64_t next_call_time;
+
+    ret = rcl_timer_get_next_call_time(timer_data->rcl_ref.get(), &next_call_time);
+
+    if (ret == RCL_RET_OK) {
+      running_timers.emplace(next_call_time, timer_data);
+    }
+
+    // wake up the timer thread so that it can pick up the timer
+    used_clock_for_timers.cancel_sleep_or_wait();
+  }
+
+  /**
+   * Returns the time when the next timer becomes ready
+   */
+  std::chrono::nanoseconds get_next_timer_ready_time() const
+  {
+    if (running_timers.empty()) {
+      // can't use std::chrono::nanoseconds::max, as wait_for
+      // internally computes end time by using ::now() + timeout
+      // as a workaround, we use some absurd high timeout
+      return std::chrono::nanoseconds(used_clock_for_timers.now().nanoseconds()) + std::chrono::hours(10000);
+    }
+    return running_timers.begin()->first;
+  }
+
+  void call_ready_timer_callbacks()
+  {
+    auto readd_timer_to_running_map = [this](TimerMap::node_type && e)
+      {
+        const auto & timer_data = e.mapped();
+        if(remove_if_dropped(timer_data))
```

Comment:
do we need this check here? It's already done at the beginning of each `while (!running_timers.empty()) {` loop iteration